File: src/host/kts/src/rvm.kt
import kotlin.system.exitProcess

typealias Rib = ArrayList<Any>
typealias GenericRib = ArrayList<*>

const val MAX_BYTE = 92
const val MULTI_BYTE_THRESHOLD = MAX_BYTE / 2
const val BYTE_OFFSET = 35
const val BYTE_DEFAULT = '9'.code      // 57
const val STRING_SEP = ','.code        // 44
const val SYMBOL_TABLE_END = ';'.code  // 59

val SHORT_ENCODINGS = arrayOf(20, 30, 0, 10, 11, 4)

//

const val CODE = ");'u?>vD?>vRD?>vRA?>vRA?>vR:?>vR=!(:lkm!':lkv6y"  // @@(replace __SOURCE__ source)@@

fun rib(x: Any, y: Any, z: Any): Rib = arrayListOf(x, y, z)

fun putChar(c: Any): Int {
    print((c as Int).toChar())
    System.out.flush() // FIXME: may be useless
    return c
}

fun getCar() {
    val char = System.`in`.readNBytes(1)[0].toInt()
    push(if (char == -1) -1 else char.toChar())
}

fun push(x: Any) {
    stack = rib(x, stack, 0)
}

fun pop(): Any {
    val valuePopped = stack.asRib()[0]
    stack = stack.asRib()[1]  // set stack to the inner stack
    return valuePopped
}

inline fun <reified T> popAs() = pop() as T

fun setGlobal(value: Any) {
    symbolTable.asRib()[0].asRib()[0] = value
    symbolTable = symbolTable.asRib()[1]
}

// @@(feature boolean
fun toBool(x: Boolean): Rib {
    return if (x) TRUE else FALSE
}
// )@@

fun getByte(): Int = CODE[pos++].code

fun getCode(): Int {
    val adjustedByte = getByte() - BYTE_OFFSET
    return if (adjustedByte < 0) BYTE_DEFAULT else adjustedByte
}

fun getInt(n: Int): Int {
    val code = getCode()
    val offset = n * MULTI_BYTE_THRESHOLD  // n represents the number of offsets x is getting
    return if (code < MULTI_BYTE_THRESHOLD) code + offset else getInt(code + offset - MULTI_BYTE_THRESHOLD)
}


inline fun prim1(crossinline f: (Any) -> Any): () -> Unit {
    return { push(f(pop())) }
}

inline fun prim2(crossinline f: (Any, Any) -> Any): () -> Unit {
    return { push(f(pop(), pop())) }
}

inline fun prim2Rib(crossinline f: (Any, Rib) -> Any): () -> Unit {
    return { push(f(pop(), popAs())) }
}

inline fun prim3(crossinline f: (Any, Any, Any) -> Any): () -> Unit {
    return { push(f(pop(), pop(), pop())) }
}

fun arg2() {
    val x = pop()
    pop()
    push(x)
}

fun close() {
    push(rib(popAs<Rib>()[0], stack, 1))
}

fun fieldSet0(y: Any, x: Rib): Any {
    x[0] = y
    return y
}

fun fieldSet1(y: Any, x: Rib): Any {
    x[1] = y
    return y
}

fun fieldSet2(y: Any, x: Rib): Any {
    x[2] = y
    return y
}

/**
 * @return The next rib that doesn't end with a 0.
 */
fun getCont(): Rib {
    var s: Rib = stack.asRib()
    while (s[2] == 0) s = s[1].asRib()
    return s
}

fun getOpND(o: Any): Any =
    if (isRib(o)) {
        o
    } else {
        listTail(stack.asRib(), o as Int)
    }


fun getSymbolFromRef(ref: Int): Any {
    return listTail(symbolTable.asRib(), ref)[0]
}

fun isRib(x: Any) = x is GenericRib

fun listTail(rib: Rib, index: Int): Rib =
    if (index == 0) {
        rib
    } else {
        listTail(rib[1].asRib(), index - 1)
    }

var PRIMITIVES = arrayOf(
    // @@(primitives (gen body ",")
    prim3 { z, y, x -> rib(x, y, z) },  // @@(primitive (rib z y x))@@
    prim1 { it },                       // @@(primitive (id x))@@
    ::pop,                              // @@(primitive (pop))@@
    ::arg2,                             // @@(primitive (arg2))@@
    ::close,                            // @@(primitive (close))@@
    prim1 { toBool(isRib(it)) },        // @@(primitive (rib? x))@@
    prim1 { it.asRib()[0] },            // @@(primitive (field-0 x))@@
    prim1 { it.asRib()[1] },            // @@(primitive (field-1 x))@@
    prim1 { it.asRib()[2] },            // @@(primitive (field-2 x))@@
    prim2Rib(::fieldSet0),              // @@(primitive (set-field-0 x))@@
    prim2Rib(::fieldSet1),              // @@(primitive (set-field-1 x))@@
    prim2Rib(::fieldSet2),              // @@(primitive (set-field-2 x))@@
    prim2 { y, x -> toBool(if (isRib(x) or isRib(y)) x === y else x == y) }, // @@(primitive (eqv? y x))@@
    prim2 { y, x -> toBool((x as Int) < y as Int) }, // @@(primitive (< x y))@@
    prim2 { y, x -> (x as Int) + y as Int }, // @@(primitive (+ x y))@@
    prim2 { y, x -> (x as Int) - y as Int }, // @@(primitive (- x y))@@
    prim2 { y, x -> (x as Int) * y as Int }, // @@(primitive (* x y))@@
    prim2 { y, x -> (x as Int) / y as Int }, // @@(primitive (/ x y))@@
    ::getCar,                            // @@(primitive (get-char))@@
    prim1(::putChar),                    // @@(primitive (put-char c))@@
    prim1 { exitProcess(0) },      // @@(primitive (exit))@@
    // )@@
)

var pos = 0
val FALSE: Rib = rib(0, 0, 5)
val TRUE: Rib = rib(0, 0, 5)
val NIL: Rib = rib(0, 0, 5)


fun buildSymbolTable(): Rib {
    var symbolTable = NIL
    var n = getInt(0)
    while (n-- > 0) symbolTable = rib(
        rib(FALSE, rib(NIL, 0, 3), 2), symbolTable, 0
    )

    var chars = NIL
    var currentStringLen = 0
    var currentByte: Int = getByte()
    while (currentByte != SYMBOL_TABLE_END) {
        if (currentByte == STRING_SEP) {
            symbolTable = rib(rib(FALSE, rib(chars, currentStringLen, 3), 2), symbolTable, 0)
            chars = NIL
            currentStringLen = 0
        } else {
            chars = rib(currentByte, chars, 0)
            currentStringLen++
        }
        currentByte = getByte()
    }
    symbolTable = rib(rib(FALSE, rib(chars, currentStringLen, 3), 2), symbolTable, 0)
    return symbolTable
}


fun stackIsValid(): Boolean {
    return ((isRib(stack) && stack.asRib().isNotEmpty())
            && (stack is Int && stack != 0))
}

fun decodeRVM(): Rib {
    var n: Any
    while (true) {
        val code = getCode()
        n = code
        var d: Int
        var op = 0
        while (true) {
            d = SHORT_ENCODINGS[op]
            if (n <= 2 + d) break
            n -= d + 3
            op++
        }
        if (code > 90) {
            n = popAs()
        } else {
            if (op == 0) {
                stack = rib(0, stack, 0)
                op++
            }
            n = when {
                n == d -> getInt(0)
                n > d -> getSymbolFromRef(getInt(n - d - 1))
                op < 3 -> getSymbolFromRef(n)
                else -> n
            }
            if (op > 4) {
                n = rib(rib(n, 0, pop()), NIL, 1)
                if (!stackIsValid()) break
                op = 4
            }
        }
        stack.asRib()[0] = rib(op - 1, n, stack.asRib()[0])
    }
    return n.asRib()[0].asRib()[2].asRib()
}

var stack: Any = NIL
var symbolTable: Any = buildSymbolTable()
var pc = decodeRVM()
fun main() {
    setGlobal(rib(0, symbolTable, 1))
    setGlobal(FALSE)
    setGlobal(TRUE)
    setGlobal(NIL)
    stack = rib(0, 0, rib(5, 0, 0))  // primordial continuation (executes the halt instruction)
    while (true) {
        var o = pc[1]
        val i = pc[0] as Int
        when (i) {
            0 -> { // jump/call
                o = getOpND(o).asRib()[0]
                var c = o.asRib()[0]
                if (isRib(c)) {
                    val c2 = rib(0, o, 0)
                    var s2 = c2
                    var nargs = c.asRib()[0] as Int
                    while (nargs > 0) {
                        s2 = rib(pop(), s2, 0)
                        nargs--
                    }
                    if (isRib(pc[2])) { // call
                        c2[0] = stack
                        c2[2] = pc[2]
                    } else { // jump
                        val k = getCont()
                        c2[0] = k[0]
                        c2[2] = k[2]
                    }
                    stack = s2
                } else {  // primitive jump/call
                    PRIMITIVES[c as Int]()
                    if (isRib(pc[2])) {  // call
                        c = pc
                    } else {  // jump
                        c = getCont()
                        stack.asRib()[1] = c[0]
                    }
                }
                pc = c.asRib()[2].asRib()
            }

            1 -> {  // set
                getOpND(o).asRib()[0] = pop()
                pc = pc[2].asRib()
            }

            2 -> {  // get
                push(getOpND(o).asRib()[0])
                pc = pc[2].asRib()
            }

            3 -> {  // const
                push(o)
                pc = pc[2].asRib()
            }

            4 -> {  // if
                pc = pc[if (pop() === FALSE) 2 else 1].asRib()
            }

            // halt (exit)
            else -> break
        }
    }
}


@Suppress("UNCHECKED_CAST")
inline fun <reified T> T.asRib() = this as Rib

File: src/host/ml/rvm.ml
(* RVM code that prints HELLO! *)
(* @@(replace ");'u?>vD?>vRD?>vRA?>vRA?>vR:?>vR=!(:lkm!':lkv6y" (encode 92) @@ *)
let input = ");'u?>vD?>vRD?>vRA?>vRA?>vR:?>vR=!(:lkm!':lkv6y"
(* )@@ *)

let debug = Sys.getenv_opt "RIBBIT_DEBUG" |> Option.is_some
let tracing = ref false
let step_count = ref 0
let start_tracing = ref 0
let next_stamp = ref 0

let _ = if debug then Printexc.record_backtrace true

module Rib = struct
  (* Rib (car, cdr, tag) *)
  type t = Rib of t ref * t ref * t ref | Integer of int

  let make_rib_of_ints a b c =
    Rib (ref (Integer a), ref (Integer b), ref (Integer c))
  let make_rib a b c = Rib (ref a, ref b, ref c)

  let is_rib = function Rib _ -> true | _ -> false
  let is_int = function Integer _ -> true | _ -> false

  let int_val = function Integer i -> i
                     | _ -> invalid_arg "int_val expects an Integer"
  let int_val_orelse i def =
    match i with
    |  Integer i -> i
    | _ -> def

  let rib_eq a b =
    match a, b with
    | Integer x, Integer y -> x = y
    | Rib _, Rib _ -> a == b
    | _ -> false

  let get_car = function Rib (car,_,_) -> !car
                       | _ -> invalid_arg "get_car expects a rib"
  let set_car rib newval =
    match rib with
    | Rib (car,_,_) -> car := newval
    | _ -> invalid_arg "set_car expects a rib"
  let get_cdr = function Rib (_,cdr,_) -> !cdr
                       | _ -> invalid_arg "get_cdr expects a rib"
  let set_cdr rib newval =
    match rib with
    | Rib (_,cdr,_) -> cdr := newval
    | _ -> invalid_arg "set_cdr expects a rib"
  let get_tag = function Rib (_,_,tag) -> !tag
                       | _ -> invalid_arg "get_tag expects a rib"
  let set_tag rib newval =
    match rib with
    | Rib (_,_,tag) -> tag := newval
    | _ -> invalid_arg "set_tag expects a rib"

  let false_rib = make_rib_of_ints 0 0 5
  let true_rib = make_rib_of_ints 0 0 5
  let nil_rib = make_rib_of_ints 0 0 5

  let to_bool = function true -> true_rib | false -> false_rib

  let write_to_buffer rib buf =
    let rec helper r =
      match r with
      | Integer i -> Printf.bprintf buf "%d" i
      | Rib _ when r == true_rib -> Buffer.add_string buf "#t"
      | Rib _ when r == false_rib -> Buffer.add_string buf "#f"
      | Rib _ when r == nil_rib -> Buffer.add_string buf "()"
      | Rib (car, cdr, tag) ->
         let obj = ref r in
         let typ = int_val_orelse !tag (-1) in
         if typ = 4 then begin
             Buffer.add_char buf '#';
             helper !car
           end else if typ = 0 then begin
             let n = ref 0 in
             Buffer.add_char buf '(';
             helper !car;
             let obj = ref !cdr in
             while is_rib !obj && int_val_orelse (get_tag !obj) (-1) = 0 do
               if !n > 4 then begin
                   Buffer.add_string buf " ...";
                   obj := nil_rib
                 end else begin
                   Buffer.add_char buf ' ';
                   helper (get_car !obj);
                   obj := get_cdr !obj;
                   incr n
                 end
             done;
             if not (!obj == nil_rib) then begin
                 Buffer.add_string buf " . ";
                 helper !obj
               end;
             Buffer.add_char buf ')'
           end else if typ = 1 then begin
             if is_rib !car then
               Printf.bprintf buf "#<procedure params=%d>" (int_val (get_car !car))
             else
               Printf.bprintf buf "#<primitive %d>" (int_val !car)
           end else if typ = 2 then begin
             obj := get_cdr !obj;
             if is_rib !obj && int_val (get_tag !obj) = 3 &&
                  int_val (get_cdr !obj) > 0 then begin
                 obj := get_car !obj;
                 while is_rib !obj && int_val (get_tag !obj) = 0 do
                   Buffer.add_char buf (get_car !obj |> int_val |> char_of_int);
                   obj := get_cdr !obj
                 done;
               end else begin
                 Buffer.add_string buf "#<symbol ";
                 helper !obj;
                 Buffer.add_char buf '>'
               end
           end else if typ = 3 then begin
             Buffer.add_char buf '"';
             obj := !car;
             while is_rib !obj && int_val (get_tag !obj) = 0 do
               begin
                 match (get_car !obj |> int_val |> char_of_int) with
                 | '\n' -> Buffer.add_string buf "\\n"
                 | '\r' -> Buffer.add_string buf "\\r"
                 | '\t' -> Buffer.add_string buf "\\t"
                 | '\\' -> Buffer.add_string buf "\\\\"
                 | '"' -> Buffer.add_string buf "\\\""
                 | c -> Buffer.add_char buf c
               end;
               obj := get_cdr !obj
             done;
             Buffer.add_char buf '"'
           end else begin
             Buffer.add_char buf '[';
             helper !car;
             Buffer.add_char buf ',';
             helper !cdr;
             Buffer.add_char  buf ',';
             helper !tag;
             Buffer.add_char buf ']'
           end in
    helper rib;
    buf

  let print_rib ?(out=stdout) rib =
    Buffer.create 16 |> write_to_buffer rib |> Buffer.output_buffer out
end

open Rib

let stack = ref (Integer 0)

let start_step () =
  incr step_count;
  if !step_count >= !start_tracing then tracing := true;
  if not !tracing then begin
      if !step_count >= !next_stamp then begin
          next_stamp := (float_of_int !next_stamp) *. 1.01 +. 1.0 |>  int_of_float;
          Printf.printf "@%d\n" !step_count;
        end
    end else begin
      let s = ref !stack in
      let buf = Buffer.create 80 in
      let sep = ref "" in
      Printf.bprintf buf "@%d STACK = (" !step_count;
      while is_rib !s && int_val_orelse (get_tag !s) (-1) = 0 do
        Buffer.add_string buf !sep;
        write_to_buffer (get_car !s) buf |> ignore;
        sep := " ";
        s := get_cdr !s
      done;
      Buffer.add_char buf ')';
      Buffer.output_buffer stdout buf;
      print_newline ();
      flush stdout
    end

let push x = stack := make_rib x !stack (Integer 0)
let pop () =
  match !stack with
  | Rib (car,cdr,_) ->
     stack := !cdr;
     !car
  | _ -> invalid_arg "Top of stack is not a rib"

module type PRIMITIVES = sig
  val primitives: (unit -> unit) array
end

module Primitives : PRIMITIVES = struct
  let prim0 f = function () -> f () |> push
  let prim1 f = function () -> pop () |> f |> push
  let prim2 f = function () ->
                  let x = pop () in
                  let y = pop () in
                  f x y |> push
  let prim3 f = function () ->
                  let x = pop () in
                  let y = pop () in
                  let z = pop () in
                  f x y z |> push

  let getchar () =
    try
      input_char stdin |> int_of_char
    with End_of_file -> -1

  let putchar c =
    char_of_int c |> print_char;
    flush stdout;
    c

  let primitives = [|
  (* @@(primitives (gen body) @@ *)
      prim3 (fun z y x -> make_rib x y z);                                    (* @@(primitive (##rib a b c))@@ *)
      prim1 (function x -> x);                                                (* @@(primitive (##id x))@@ *)
      (function () -> pop () |> ignore);                                      (* @@(primitive (##arg1 x y))@@ *)
      prim2 (fun y x -> y);                                                   (* @@(primitive (##arg2 x y))@@ *)
      prim1 (function x -> make_rib (get_car x) !stack (Integer 1));          (* @@(primitive (##close rib))@@ *)
      prim1 (function Rib _ -> true_rib | _ -> false_rib);                    (* @@(primitive (##rib? rib) (use bool2scm))@@ *)
      prim1 get_car;                                                          (* @@(primitive (##field0 rib))@@ *)
      prim1 get_cdr;                                                          (* @@(primitive (##field1 rib))@@ *)
      prim1 get_tag;                                                          (* @@(primitive (##field2 rib))@@ *)
      prim2 (fun y x -> set_car x y; y);                                      (* @@(primitive (##field0-set! rib))@@ *)
      prim2 (fun y x -> set_cdr x y; y);                                      (* @@(primitive (##field1-set! rib))@@ *)
      prim2 (fun y x -> set_tag x y; y);                                      (* @@(primitive (##field2-set! rib))@@ *)
      prim2 (fun y x -> to_bool (rib_eq x y));                                (* @@(primitive (##eqv? x y) (use bool2scm))@@ *)
      (*  @@(primitive (##< x y) @@ *)
      prim2 (fun y x -> match x, y with
                        | Integer a, Integer b -> to_bool (a < b)
                        | _ -> invalid_arg "< arguments must be Integers");
      (*  )@@ *)
      (*  @@(primitive (##+ x y) @@ *)
      prim2 (fun y x -> match x, y with
                        | Integer a, Integer b -> Integer (a + b)
                        | _ -> invalid_arg "+ arguments must be Integers");
      (*  )@@ *)
      (*  @@(primitive (##- x y) @@ *)
      prim2 (fun y x -> match x, y with
                        | Integer a, Integer b -> Integer (a - b)
                        | _ -> invalid_arg "- arguments must be Integers");
      (*  )@@ *)
      (*  @@(primitive (##* x y) @@ *)
      prim2 (fun y x -> match x, y with
                        | Integer a, Integer b -> Integer (a * b)
                        | _ -> invalid_arg "* arguments must be Integers");
      (*  )@@ *)
      (*  @@(primitive (##quotient x y) @@ *)
      prim2 (fun y x -> match x, y with
                        | Integer a, Integer b -> Integer (a / b)
                        | _ -> invalid_arg "quotient arguments must be Integers");
      (*  )@@ *)
      prim0 (function () -> Integer (getchar ()));                                                                (* @@(primitive (##getchar))@@ *)
      prim1 (function Integer ch -> Integer (putchar ch) | _ -> invalid_arg "putchar argument must be Integer");  (* @@(primitive (##putchar x))@@ *)
      prim1 (function Integer status -> exit status | _ -> invalid_arg "exit argument must be Integer")           (* @@(primitive (##exit x))@@ *)
  (*  )@@ *)
    |]
end

let byte_count = ref 0

let get_byte = function () ->
  let c = input.[!byte_count] in
  incr byte_count;
  int_of_char c

let get_code () =
  let x = (get_byte ()) - 35 in if x < 0 then 57 else x

let rec get_int n =
  let x = get_code () in
  let n = n * 46 in
  if x < 46 then n + x else get_int (n + x - 46)

let rec list_tail lst i =
  if i = 0 then lst else list_tail (get_cdr lst) (i - 1)

(* Build the initial symbol table *)
let symtbl = ref nil_rib
let _ =
  for n = get_int 0 downto 1 do
    symtbl := make_rib (make_rib false_rib
                          (make_rib nil_rib (Integer 0) (Integer 3))
                          (Integer 2))
                !symtbl (Integer 0)
  done;
  let accum = ref nil_rib
  and n = ref 0
  and in_loop = ref true in
  while !in_loop do
    let c = get_byte () in
    if c = 44 then begin
        let r1 = make_rib !accum (Integer !n) (Integer 3) in
        let r2 = make_rib false_rib r1 (Integer 2) in
        symtbl := make_rib r2 !symtbl (Integer 0);
        accum := nil_rib;
        n := 0
      end
    else if c = 59 then
      in_loop := false
    else begin
        accum := make_rib (Integer c) !accum (Integer 0);
        incr n
      end
  done;
  symtbl := make_rib (make_rib false_rib
                        (make_rib !accum (Integer !n) (Integer 3))
                        (Integer 2))
              !symtbl (Integer 0)

let symbol_ref n = list_tail !symtbl n |> get_car

let pc = ref nil_rib

(* Decode the RVM instructions *)
let _ =
  let codes = [| 20; 30; 0; 10; 11; 4 |]
  and in_loop = ref true
  and n = ref nil_rib in
  while !in_loop do
    let x = get_code ()
    and d = ref 0
    and op = ref 0
    and in_loop2 = ref true in
    n := Integer x;
    while !in_loop2 do
      d := codes.(!op);
      if int_val !n <= 2 + !d then
        in_loop2 := false
      else begin
          n := Integer ((int_val !n) - (!d + 3));
          incr op
        end
    done;
    if x > 90 then
      n := pop ()
    else begin
        if !op = 0 then begin
            stack := make_rib (Integer 0) !stack (Integer 0);
            incr op
          end;
        n := if int_val !n = !d then
               Integer (get_int 0)
             else if int_val !n >= !d then
               get_int ((int_val !n) - !d - 1) |> symbol_ref
             else if !op < 3 then
               int_val !n |> symbol_ref
             else
               !n;
        if 4 < !op then begin
            n := make_rib (make_rib !n (Integer 0) (pop ())) nil_rib (Integer 1);
            if not (is_rib !stack) then
              in_loop := false
            else
              op := 4
          end;
      end;
    if !in_loop then
      set_car !stack (make_rib (Integer (!op - 1)) !n (get_car !stack))
  done;
  pc := get_car !n |> get_tag

let get_opnd o =
  match o with
  | Rib _ -> o
  | Integer i -> list_tail !stack i

let get_cont () =
  let s = ref !stack in
  while not (is_rib (get_tag !s)) do
    s := get_cdr !s
  done;
  !s

let set_global v =
  set_car (get_car !symtbl) v;
  symtbl := (get_cdr !symtbl)

(* Needed for shift_right function *)
open Int
(* Execute the program *)
let _ =
  set_global (make_rib (Integer 0) !symtbl (Integer 1));
  set_global false_rib;
  set_global true_rib;
  set_global nil_rib;
  stack := make_rib (Integer 0) (Integer 0) (make_rib_of_ints 5 0 0);
  while true do
    if debug then start_step ();
    let o = ref (get_cdr !pc)
    and i = int_val (get_car !pc) in
    if i < 1 then begin (* jump/call *)
        if !tracing then begin
            (if is_rib (get_tag !pc) then "call " else "jump ") |> prerr_string;
            print_rib ~out:stderr !o;
            prerr_newline ()
          end;
        o := get_opnd !o |> get_car;
        let c = ref (get_car !o) in
        if is_rib !c then begin
            let c2 = ref (make_rib (Integer 0) !o (Integer 0)) in
            let s2 = ref !c2
            and nargs = shift_right (int_val (get_car !c)) 1 in
            for narg = nargs downto 1 do
              s2 := make_rib (pop ()) !s2 (Integer 0)
            done;
            if is_rib (get_tag !pc) then begin (* call *)
                set_car !c2 !stack;
                set_tag !c2 (get_tag !pc)
              end else begin (* jump *)
                let k = get_cont () in
                set_car !c2 (get_car k);
                set_tag !c2 (get_tag k)
              end;
            stack := !s2
          end else begin
            Primitives.primitives.(int_val !c) ();
            if is_rib (get_tag !pc) then (* call *)
              c := !pc
            else begin (* jump *)
                c := get_cont ();
                set_cdr !stack (get_car !c)
              end
          end;
        pc := get_tag !c
      end else if i < 2 then begin (* set *)
        if !tracing then begin
            prerr_string "set ";
            print_rib ~out:stderr !o;
            prerr_newline ()
          end;
        let opnd = get_opnd !o in
        set_car opnd (get_car !stack);
        stack := get_cdr !stack;
        pc := get_tag !pc
      end else if i < 3 then begin (* get *)
        if !tracing then begin
            prerr_string "get ";
            print_rib ~out:stderr !o;
            prerr_newline ()
          end;
        let opnd = get_opnd !o in
        push (get_car opnd);
        pc := get_tag !pc
      end else if i < 4 then begin (* const *)
        if !tracing then begin
            prerr_string "const ";
            print_rib ~out:stderr !o;
            prerr_newline ()
          end;
        push !o;
        pc := get_tag !pc
      end else if i < 5 then begin (* if *)
        if !tracing then prerr_endline "if";
        pc := if pop () == false_rib then get_tag !pc else get_cdr !pc
      end else begin (* halt *)
        if !tracing then prerr_endline "halt";
        exit 0
      end
  done
